sympy/core/basic.py

```diff
@@ -1624,31 +1624,73 @@ def count_ops(self, visual=None):
         return count_ops(self, visual)
 
     def doit(self, **hints):
-        """Evaluate objects that are not evaluated by default like limits,
-        integrals, sums and products. All objects of this kind will be
-        evaluated recursively, unless some species were excluded via 'hints'
-        or unless the 'deep' hint was set to 'False'.
+        """Evaluate objects that are not evaluated like operations, limits,
+        integrals, sums and products.
 
-        >>> from sympy import Integral
+        Explanation
+        ===========
+
+        Arguments will be evaluated recursively, unless some species were
+        excluded via 'properties' or unless the 'deep' hint was set to 'False'.
+        Evaluatable classes should define their own `_eval_derivative` method,
+        which returns evaluated result. In this case, non-recursive evaluation
+        should also be done when `deep=False` is passed. It is also encouraged
+        to allow `evaluate` option to the constructor.
+
+        Examples
+        ========
+
+        >>> from sympy import Add, Integral, Derivative
         >>> from sympy.abc import x
 
-        >>> 2*Integral(x, x)
-        2*Integral(x, x)
+        >>> expr = Add(Integral(x, x), Integral(x, x), Derivative(x, x), evaluate=False)
+        >>> expr
+        Derivative(x, x) + Integral(x, x) + Integral(x, x)
+
+        >>> expr.doit()
+        x**2 + 1
 
-        >>> (2*Integral(x, x)).doit()
-        x**2
+        >>> expr.doit(deep=False)
+        Derivative(x, x) + 2*Integral(x, x)
 
-        >>> (2*Integral(x, x)).doit(deep=False)
-        2*Integral(x, x)
+        See Also
+        ========
 
+        core.operations.AssocOp
         """
-        if hints.get('deep', True):
-            terms = [term.doit(**hints) if isinstance(term, Basic) else term
-                                         for term in self.args]
+        deep = hints.get('deep', True)
+        properties = hints.get('properties', [])
+
+        eval_doit = self._eval_doit(**hints)
+        if eval_doit is not None:
+            return eval_doit
+        if deep:
```

Comment:
If deep is true then we should have the args doit before calling `_eval_doit` here.

Reply:
I don't want to restrict the ability of `_eval_doit` here. Some classes may want to do some processings with unevaluated `self.args` before evaluating themselves.

Reply:
Oh, on second thought, I think you're right.